app/src/org/commcare/tasks/UnzipTask.java

```diff
@@ -1,109 +1,137 @@
+    }
 
-            if (entry.isDirectory()) {
-                FileUtil.createFolder(new File(destination, entry.getName()).toString());
-                //If it's a directory we can move on to the next one
-                continue;
-            }
+    private Integer unZipFromStream(ZipInputStream zis, String destinationPath) {
+        File destination = new File(destinationPath);
+        int count = 0;
+        ZipEntry entry = null;
+        try {
+            while ((entry = zis.getNextEntry()) != null) {
+                publishProgress(Localization.get("mult.install.progress", new String[]{String.valueOf(count)}));
+                count++;
 
-            File outputFile = new File(destination, entry.getName());
-            if (!outputFile.getParentFile().exists()) {
-                FileUtil.createFolder(outputFile.getParentFile().toString());
-            }
-            if (outputFile.exists()) {
-                //Try to overwrite if we can
-                if (!outputFile.delete()) {
-                    //If we couldn't, just skip for now
+                if (entry.isDirectory()) {
+                    FileUtil.createFolder(new File(destination, entry.getName()).toString());
+                    //If it's a directory we can move on to the next one
                     continue;
                 }
-            }
-            BufferedInputStream inputStream;
-            try {
-                inputStream = new BufferedInputStream(zipfile.getInputStream(entry));
-            } catch (IOException ioe) {
-                this.publishProgress(Localization.get("mult.install.progress.badentry", new String[]{entry.getName()}));
-                return -1;
-            }
-
-            BufferedOutputStream outputStream;
-            try {
-                outputStream = new BufferedOutputStream(new FileOutputStream(outputFile));
-            } catch (IOException ioe) {
-                this.publishProgress(Localization.get("mult.install.progress.baddest", new String[]{outputFile.getName()}));
-                return -1;
-            }
 
-            try {
-                try {
-                    StreamsUtil.writeFromInputToOutputNew(inputStream, outputStream);
-                } catch (IOException ioe) {
-                    this.publishProgress(Localization.get("mult.install.progress.errormoving"));
-                    return -1;
+                File outputFile = new File(destination, entry.getName());
+                if (!outputFile.getParentFile().exists()) {
+                    FileUtil.createFolder(outputFile.getParentFile().toString());
                 }
-            } finally {
-                try {
-                    outputStream.close();
-                } catch (IOException ioe) {
+                if (outputFile.exists()) {
+                    //Try to overwrite if we can
+                    if (!outputFile.delete()) {
+                        //If we couldn't, just skip for now
+                        continue;

```

Comment:
Why is this OK? Couldn't this potentially result in an entry from the zip file not being copied (and then lost once the zip is deleted) silently?

Reply:
I am not sure as well, I took it from master equivalent over here - https://github.com/dimagi/commcare-android/blob/master/app/src/org/commcare/tasks/UnzipTask.java#L67 . Maybe @ctsims know more about why are we doing this ?

Reply:
Given that this is inherited doesn't need to block, but we should throw an Exception here and halt the process